# ITMGT 25.03 EE
# Assignment 3B

This is a data-cleaning assignment and is the second of two parts. This notebook is uploaded to our Github repository along with a csv of fake, dirty data called **recweekdata.csv**.

In addition, you have a new file called **studentprofiles.csv** which has first and last names plus the genders of the students.

The output of this assignment is to convert recweekdata.csv into processable data for statistics. Submit the following files through GitHub:
1. A cleaned recweekdata.csv called **cleanrecweekdata3b.csv** **(5 points)**
2. This notebook, containing the code (properly commented) that you used to clean the data **(5 points)**
3. An html file of summary statistics from each column called **stats3b.html**. The requirements for the summary statistics are described below. **(10 points)**

The outputs are similar to that of Assignment 3A, except that you will now use Pandas instead of Python lists.

This entire assignment (3B) is worth 20 points.

Upload your output files to your Github repositories. Make sure to send an email to **joben.ilagan@gmail.com** once you're done.

This assignment is due on **Monday, July 9, 2019** at 12:00NN.

Note: I understand you all need to review for LT 2. Think of this Assignment as practice. It's not meant to take you as long as 3A.

#### Bonus Points:
* +2 if submitted over (or before) the weekend (July 13-14)
* +2 if CSS and div tags are used (but please don't spend too much time on this because you still need to review for the Long Test)

In [2]:
## Code goes here
import pandas as pd
import csv
import re



In [7]:
#Making lists that turn into the columns
Firstnames= []
Lastnames=[]
Dirty_Birthmonths=[]
Birthdays=[]
Dirty_Birthyears=[]
Dirty_Courses=[]
Dirty_Phonenumbers=[]

with open('recweekdata.txt') as csv_file:
    csv_reader = csv.reader(csv_file, delimiter=',')
    line_count = 0
    for row in csv_reader:
        if line_count == 0:
            print(f'Column names are {", ".join(row)}')
            line_count += 1
        else:
            Firstnames.append(row[0])
            Lastnames.append(row[1])
            Dirty_Birthmonths.append(row[2])
            Birthdays.append(row[3])
            Dirty_Birthyears.append(row[4])
            Dirty_Courses.append(row[5])
            Dirty_Phonenumbers.append(row[6])
            line_count += 1
            
    print(f'Processed {line_count} lines.')



Column names are firstname, lastname, birthmonth, birthday, birthyear, course, phonenumber
Processed 3001 lines.


In [10]:
#turning the genders into a list so that I don't need to join two dataframes 
Genders=[]             
with open('studentprofiles.txt') as csv_file:
    csv_reader = csv.reader(csv_file, delimiter=',')
    line_count = 0
    for row in csv_reader:
        if line_count == 0:
            print(f'Column names are {", ".join(row)}')
            line_count += 1
        else:
            Genders.append(row[2])
            line_count += 1

Column names are firstname, lastname, gender


In [90]:
df = pd.DataFrame({"firstname":Firstnames, "lastname":Lastnames, "birthmonth":Dirty_Birthmonths, 
                   "birthday": Birthdays, "birthyear": Dirty_Birthyears, "course":Dirty_Courses,"phonenumber":Dirty_Phonenumbers,
                  "gender":Genders})
df

firstname      lastname birthmonth birthday birthyear  \
0        name1     lastname1         11       26      1998   
1        name2     lastname2        May       26      2001   
2        name3     lastname3          5       16      1997   
3        name4     lastname4        Dec        9        97   
4        name5     lastname5        Sep        5        97   
5        name6     lastname6      March        2      1999   
6        name7     lastname7         11       10        98   
7        name8     lastname8        Aug       28        99   
8        name9     lastname9       July       10        97   
9       name10    lastname10          9       14      2001   
10      name11    lastname11   December        8        98   
11      name12    lastname12          5        4        97   
12      name13    lastname13    January       28      1999   
13      name14    lastname14          1       11      1999   
14      name15    lastname15          6        6        98   
15      name16    lastname16          4       20        99   
16      name17    lastname17          1        3        97   
17      name18    lastname18        Apr        8      1997   
18      name19    lastname19         12       28      1997   
19      name20    lastname20          3       10        97   
20      name21    lastname21          7       13        99   
21      name22    lastname22          9       15      1999   
22      name23    lastname23          4       15        97   
23      name24    lastname24        Oct        1        97   
24      name25    lastname25          5        6        97   
25      name26    lastname26         12        5      1999   
26      name27    lastname27      March        4      1997   
27      name28    lastname28         11       10      2000   
28      name29    lastname29         10       10      1999   
29      name30    lastname30        May       28      1999   
...        ...           ...        ...      ...       ...   
2970  name2971  lastname2971   February       25        97   
2971  name2972  lastname2972        May       14      1998   
2972  name2973  lastname2973        Oct        9      1999   
2973  name2974  lastname2974          8       12      1999   
2974  name2975  lastname2975   November       18      1997   
2975  name2976  lastname2976       July       24      1998   
2976  name2977  lastname2977        May       26        98   
2977  name2978  lastname2978         10       26        99   
2978  name2979  lastname2979          2       13      1997   
2979  name2980  lastname2980       July        3        99   
2980  name2981  lastname2981        Feb       17      1998   
2981  name2982  lastname2982          3       28      2000   
2982  name2983  lastname2983   December        3        98   
2983  name2984  lastname2984        Nov       13      2000   
2984  name2985  lastname2985        Feb       23        99   
2985  name2986  lastname2986          3        2      1999   
2986  name2987  lastname2987        Oct        6        97   
2987  name2988  lastname2988        Feb       20      1997   
2988  name2989  lastname2989        Jan       17      1999   
2989  name2990  lastname2990        Jul       14        97   
2990  name2991  lastname2991       July       27      2000   
2991  name2992  lastname2992      March       26        99   
2992  name2993  lastname2993          4        1      2000   
2993  name2994  lastname2994      April       10        99   
2994  name2995  lastname2995        May       21        97   
2995  name2996  lastname2996      March       28        97   
2996  name2997  lastname2997        May       19      1998   
2997  name2998  lastname2998       July        3        98   
2998  name2999  lastname2999         12        2        98   
2999  name3000  lastname3000        Jul       13      1997   

                                       course       phonenumber  gender  
0                                       BS LM  +63 918 302 6210  Female  
1           

In [91]:
#Brute forcing replacements
df.loc[:,"phonenumber"].replace(r"^\+63","",regex=True, inplace=True)
df.loc[:,"phonenumber"].replace(r"^ ","",regex=True, inplace=True)
carrier = {"917":"Globe","918":"Smart", "927":"Globe"}

df["carrier"] = df.loc[:,"phonenumber"].str[0:3].apply(lambda x: carrier[x])
df.loc[:,"birthyear"].replace(r"^97","1997",regex=True, inplace = True)
df.loc[:,"birthyear"].replace(r"^98","1998",regex=True, inplace = True)
df.loc[:,"birthyear"].replace(r"^99","1999",regex=True, inplace = True)

df.loc[:,"course"].replace(r"^Management Engineering","BS ME",regex=True, inplace = True)
df.loc[:,"course"].replace(r"^Management","BS MGT",regex=True, inplace = True)
df.loc[:,"course"].replace(r"^ME","BS ME",regex=True, inplace = True)
df.loc[:,"course"].replace(r"^ITE","BS ITE",regex=True, inplace = True)
df.loc[:,"course"].replace(r"^CTM","BS CTM",regex=True, inplace = True)
df.loc[:,"course"].replace(r"^LM","BS LM",regex=True, inplace = True)
df.loc[:,"course"].replace(r"^MGT","BS MGT",regex=True, inplace = True)
df.loc[:,"course"].replace(r"^Information Technology Entrepreneurship","BS ITE",regex=True, inplace = True)
df.loc[:,"course"].replace(r"^Communications Technology","BS CTM",regex=True, inplace = True)
df.loc[:,"course"].replace(r"^Legal Management","BS LM",regex=True, inplace = True)

df.loc[:,"birthmonth"].replace(r"^10","October",regex=True, inplace = True)
df.loc[:,"birthmonth"].replace(r"^11","November",regex=True, inplace = True)
df.loc[:,"birthmonth"].replace(r"^12","December",regex=True, inplace = True)
df.loc[:,"birthmonth"].replace(r"^1","January",regex=True, inplace = True)
df.loc[:,"birthmonth"].replace(r"^2","February",regex=True, inplace = True)
df.loc[:,"birthmonth"].replace(r"^3","March",regex=True, inplace = True)
df.loc[:,"birthmonth"].replace(r"^4","April",regex=True, inplace = True)
df.loc[:,"birthmonth"].replace(r"^5","May",regex=True, inplace = True)
df.loc[:,"birthmonth"].replace(r"^6","June",regex=True, inplace = True)
df.loc[:,"birthmonth"].replace(r"^7","July",regex=True, inplace = True)
df.loc[:,"birthmonth"].replace(r"^8","August",regex=True, inplace = True)
df.loc[:,"birthmonth"].replace(r"^9","September",regex=True, inplace = True)

df.loc[:,"birthmonth"].replace(r"^Julyy","July",regex=True, inplace = True)
df["birthmonth"]= df.loc[:,"birthmonth"].str[0:3]


firstname      lastname birthmonth birthday birthyear  course  \
0        name1     lastname1        Nov       26      1998   BS LM   
1        name2     lastname2        May       26      2001  BS MGT   
2        name3     lastname3        May       16      1997  BS CTM   
3        name4     lastname4        Dec        9      1997  BS ITE   
4        name5     lastname5        Sep        5      1997  BS CTM   
5        name6     lastname6        Mar        2      1999  BS MGT   
6        name7     lastname7        Nov       10      1998   BS LM   
7        name8     lastname8        Aug       28      1999   BS ME   
8        name9     lastname9        Jul       10      1997   BS ME   
9       name10    lastname10        Sep       14      2001  BS CTM   
10      name11    lastname11        Dec        8      1998   BS LM   
11      name12    lastname12        May        4      1997   BS ME   
12      name13    lastname13        Jan       28      1999  BS CTM   
13      name14    lastname14        Jan       11      1999  BS MGT   
14      name15    lastname15        Jun        6      1998  BS ITE   
15      name16    lastname16        Apr       20      1999   BS LM   
16      name17    lastname17        Jan        3      1997  BS ITE   
17      name18    lastname18        Apr        8      1997  BS MGT   
18      name19    lastname19        Dec       28      1997  BS ITE   
19      name20    lastname20        Mar       10      1997   BS ME   
20      name21    lastname21        Jul       13      1999   BS LM   
21      name22    lastname22        Sep       15      1999  BS ITE   
22      name23    lastname23        Apr       15      1997  BS MGT   
23      name24    lastname24        Oct        1      1997   BS ME   
24      name25    lastname25        May        6      1997   BS ME   
25      name26    lastname26        Dec        5      1999   BS ME   
26      name27    lastname27        Mar        4      1997  BS CTM   
27      name28    lastname28        Nov       10      2000   BS LM   
28      name29    lastname29        Oct       10      1999   BS LM   
29      name30    lastname30        May       28      1999  BS CTM   
...        ...           ...        ...      ...       ...     ...   
2970  name2971  lastname2971        Feb       25      1997  BS MGT   
2971  name2972  lastname2972        May       14      1998   BS ME   
2972  name2973  lastname2973        Oct        9      1999   BS LM   
2973  name2974  lastname2974        Aug       12      1999  BS ITE   
2974  name2975  lastname2975        Nov       18      1997   BS LM   
2975  name2976  lastname2976        Jul       24      1998  BS ITE   
2976  name2977  lastname2977        May       26      1998   BS ME   
2977  name2978  lastname2978        Oct       26      1999  BS ITE   
2978  name2979  lastname2979        Feb       13      1997  BS CTM   
2979  name2980  lastname2980        Jul        3      1999   BS LM   
2980  name2981  lastname2981        Feb       17      1998  BS ITE   
2981  name2982  lastname2982        Mar       28      2000  BS ITE   
2982  name2983  lastname2983        Dec        3      1998  BS CTM   
2983  name2984  lastname2984        Nov       13      2000  BS MGT   
2984  name2985  lastname2985        Feb       23      1999  BS MGT   
2985  name2986  lastname2986        Mar        2      1999  BS ITE   
2986  name2987  lastname2987        Oct        6      1997  BS ITE   
2987  name2988  lastname2988        Feb       20      1997  BS MGT   
2988  name2989  lastname2989        Jan       17      1999   BS ME   
2989  name2990  lastname2990        Jul       14      1997  BS MGT   
2990  name2991  lastname2991        Jul       27      2000  BS ITE   
2991  name2992  lastname2992        Mar       26      1999  BS ITE   
2992  name2993  lastname2993        Apr        1      2000   BS ME   
2993  name2994  lastname2994        Apr       10      1999  BS CTM   
2994  name2995  lastname2995        May       21      1997  BS CTM   
2995  name2996 

In [94]:
df.to_csv("cleanrecweekdata3b.csv")

# Required Summary Statistics

#### These were the questions from 3A:

1. How many people were born in each month?
2. How many people were born on each day?
3. How many people were born in each year?
4. How many people come from each course?
5. How many people have numbers from Globe (0917 or 0927) and Smart (0918)? *Note: it's up to you how to format this output*

#### Additional Questions:
* How many Globe subscribers are: Male? Female?
* How many Smart subscribers are: Male? Female?

Answer these questions in an html file called **stats3b.html**. The format of the html is as follows (in template form, so you can use this as format string in your program, though feel free to modify with variable names if you find it easier to do so):


```
<html>
    <body>
         <h2>Summary Statistics</h2>
         <table>
             <tr><th>Stat</th><th>Value</th></tr>
             <tr><td>People born in each month</td><td>{}</td></tr>
             <tr><td>People born on each day</td><td>{}</td></tr>
             <tr><td>People born in each year</td><td>{}</td></tr>
             <tr><td>People coming from each course</td><td>{}</td></tr>
             <tr><td>People having numbers from Globe and Smart</td><td>{}</td></tr>
             <tr><td>Male Globe subscribers:</td><td>{}</td></tr>
             <tr><td>Female Globe subscribers:</td><td>{}</td></tr>
             <tr><td>Male Smart subscribers:</td><td>{}</td></tr>
             <tr><td>Female Smart subscribers:</td><td>{}</td></tr>
         </table>
    </body>
</html>
```

In [74]:
df.loc[:,["firstname","gender","carrier"]].groupby(["carrier","gender"]).count()

firstname
carrier gender           
Globe   Female        984
        Male         1061
Smart   Female        494
        Male          461

In [75]:
df.loc[:,["firstname","carrier"]].groupby(["carrier"]).count()

firstname
carrier           
Globe         2045
Smart          955

In [77]:
df.loc[:,["firstname","birthday"]].groupby(["birthday"]).count()

firstname
birthday           
1               112
10              111
11              105
12               98
13              119
14              102
15              121
16              115
17              116
18              107
19              113
2               106
20               91
21              123
22              103
23              116
24              108
25              116
26              116
27              102
28              104
3                90
4               110
5               100
6                99
7                88
8               111
9                98

In [118]:
df.loc[:,["firstname","course"]].groupby(["course"]).count()

firstname
course           
BS CTM        589
BS ITE        590
BS LM         585
BS ME         607
BS MGT        629

In [78]:
df.loc[:,["firstname","birthmonth"]].groupby(["birthmonth"]).count()

firstname
birthmonth           
Apr               267
Aug               241
Dec               276
Feb               232
Jan               270
Jul               249
Jun               259
Mar               246
May               273
Nov               277
Oct               178
Sep               232

In [86]:
df.loc[:,["firstname","birthyear"]].groupby(["birthyear"]).count()

firstname
birthyear           
1997             756
1998             746
1999             756
2000             380
2001             362

In [87]:
df.loc[df.loc[:,"gender"]=="Male",["gender"]].count()

gender    1522
dtype: int64

In [76]:
html = """
<html>
    <body>
         <h2>Summary Statistics</h2>
         <table>
             <tr><th>Stat</th><th>Value</th></tr>
             <tr><td>People born in each month</td><td>{}</td></tr>
             <tr><td>People born on each day</td><td>{}</td></tr>
             <tr><td>People born in each year</td><td>{}</td></tr>
             <tr><td>People coming from each course</td><td>{}</td></tr>
             <tr><td>People having numbers from Globe and Smart</td><td>{}</td></tr>
             <tr><td>Male Globe subscribers:</td><td>{}</td></tr>
             <tr><td>Female Globe subscribers:</td><td>{}</td></tr>
             <tr><td>Male Smart subscribers:</td><td>{}</td></tr>
             <tr><td>Female Smart subscribers:</td><td>{}</td></tr>
         </table>
    </body>
</html>
"""


In [115]:
apple = df.loc[:,["firstname","birthyear"]].groupby(["birthyear"]).count()

In [126]:
statdf= pd.DataFrame({"Stat":["Values","cont","cont","cont","cont"], "People born in each month ":["Apr:267, Aug:,241 Dec:276, Feb:232, Jan:270", "Jul:249, Jun:259, Mar:246, May:273, Nov:277", "Oct:178, Sep:23","",""],
                    "People born on each day ":["1:112, 2:106,3:90, 4:110, 5:100, 6:99, 7:88", "8:111, 9:98, 10:111, 11:105, 12:98, 13:119", "14:102, 15:121, 16:115, 17:116, 18:107, 19:113", "20:91, 21:123, 22:103, 23:116, 24:108, 25:116, 26:116", "27:102, 28:104"],
                   "People born on each year": ["1997: 756, 1998: 746, 1999: 756", "2000: 380, 2001: 362","","",""], 
                    "People coming from each course": ["BS CTM:589", "BS ITE:590", "BS LM:585", "BS ME:607", "BS MGT:629"], 
                     "People having numbers from Globe and Smart":["Globe:2045", "Smart:955","","",""],
                      "Male Globe Subscribers":["1061","","","",""], "Female Globe Subscribers":["984","","","",""],
                     "Male Smart Subscribers":["461","","","",""], "Female Smart Subscribers":["494","","","",""]})
statdf.to_html("stats3b.html")

In [120]:
df.loc[:,["firstname","gender","carrier"]].groupby(["carrier","gender"]).count()

firstname
carrier gender           
Globe   Female        984
        Male         1061
Smart   Female        494
        Male          461